### ISSUE TITLE
wasm_bindgen(skip) in cfg_attr

### ISSUE THREAD

**Issue by mBornand** [rustwasm/wasm-bindgen#2703]
Hello
I tried to use `#[cfg_attr(feature = "wasm-bindgen", wasm_bindgen(skip))]` but I get this:

```   |
99 |     #[cfg_attr(feature = "wasm-bindgen", wasm_bindgen(skip))]
   |                                          ^^^^^^^^^^^^ not a non-macro attribute
```

But it works for other macros like derive. Is this a bug or does a workaround exist?

**Comment by mBornand** [comment]
I found a solution.
```
#![feature(cfg_eval)]
use wasm_bindgen::prelude::*;

#[cfg_eval]
#[cfg_attr(feature = "wasm-bindgen", wasm_bindgen)]
pub struct Test2 {
    #[cfg_attr(feature = "wasm-bindgen", wasm_bindgen(skip))]
    pub s: Option<String>,
}
```
It works. But in some way it should be the responsibility of the `wasm_bindgen` macro to do this.

**Comment by FabianHummel** [comment]
I also have this issue but have not found a fix for it yet. When I try the supposed solution from above I just get syntax errors about expected symbols in this `#![feature(cfg_eval)]` line.

```rust
#![feature(cfg_eval)] // syntax error here
use wasm_bindgen::prelude::*;

#[cfg_eval]
#[cfg_attr(feature = "wasm-bindgen", wasm_bindgen)]
pub struct Test2 {
    #[cfg_attr(feature = "wasm-bindgen", wasm_bindgen(skip))]
    pub s: Option<String>,
}
```

```rust
#[cfg_attr(target_arch = "wasm32", wasm_bindgen::prelude::wasm_bindgen)]
pub struct MyStruct {
    #[cfg_attr(target_arch = "wasm32", wasm_bindgen::prelude::wasm_bindgen(skip))]
    pub my_field: String,
```

<img width="782" alt="image" src="https://github.com/user-attachments/assets/c102f3c6-2a9b-4c00-a9d5-0aad138a94e6">
<img width="851" alt="image" src="https://github.com/user-attachments/assets/86d6dde5-4081-4dcf-a818-d37c62197634">

**Comment by daxpedda** [comment]
I do not believe that there is something `wasm-bindgen` can do here. While it is possible for `wasm-bindgen` to parse `#[cfg_attr(...)]` attributes, we can't evaluate them, this is just a limitation of how proc-macros work in Rust.

I'm unfamiliar with `cfg_eval`, so I can't assist there.

**Comment by FabianHummel** [comment]
it's fine. i guess I'll just wait and see what the future has to offer. In the meantime, this is my workaround:

```rust
#[cfg_attr(target_arch = "wasm32", wasm_bindgen::prelude::wasm_bindgen)]
pub struct MyStruct {
    #[cfg(target_arch = "wasm32")]
    #[wasm_bindgen::prelude::wasm_bindgen(skip)]
    pub my_field: String,
    #[cfg(not(target_arch = "wasm32"))]
    pub my_field: String,
}
```

a bit cumbersome, but it works.

**Comment by fl0rek** [comment]
I believe I found a fix which doesn't require `cfg_eval`, thus can be used in stable (which motivated me to look into it). It uses similar two-step expansion pattern used to  allow `cfg_attr` in `impl` blocks: https://github.com/rustwasm/wasm-bindgen/pull/1208. It is also inspired by the discussion on a similar problem in `bon` crate: https://github.com/elastio/bon/discussions/124.

Overview:
- During `#[wasm_bindgen]` attribute macro expansion on struct, instead of generating ast for Program, it puts a `#[derive(BindgenedStruct)]` on it and preserves all the `wasm_bindgen` atributes as-is.
- When `#[derive(BindgenedStruct)]` is expanded, `cfg_attr` are already evaluated, so macro can just processes the leftover attributes and generate the glue code required, as it's done now. Attributes are then automatically removed by the compiler, since macro is defined as `#[proc_macro_derive(BindgenedStruct, attributes(wasm_bindgen))]`.

`BindgenedStruct` is a placeholder, internal-use name.

I think I have a POC working with test passing and would be happy to work on getting it to a PR state.

**Comment by daxpedda** [comment]
That sounds workable indeed. I'm hoping the implementation is actually lightweight and is covered by enough tests.
Happy to review a PR!
